docs(dropdown): usage content updates (#1212)
carbon-design-system/carbon-website: src/pages/components/dropdown/usage.mdx

---
title: Dropdown

> description: Dropdowns present a list of options that can be used to filter or sort existing content.

Author: I like that Josh! I think it encompasses all the functionalities a dropdown can have. Will add 👍🏻

## Overview

> There are three different types of dropdowns that support various kinds of functionality—dropdown, multiselect, and combobox.

Reviewer: are we supposed to use a different type of dash here? seems strange to read "...functionality-dropdown..."

Author: Yes, this Jan did a content review on that.

### Types

| Type                        | Purpose                                                                                                  |
| --------------------------- | -------------------------------------------------------------------------------------------------------- |
| [Dropdown](#dropdown)       | Allows the user to select one option from a list.                                                        |
| [Multiselect](#multiselect) | Allows the user to select multiple options from a list and filter.                                       |
| [Combobox](#combobox)       | Allows a user to enter any value, or select a value from a list of suggested, likely, or desired values. |

#### When to use

- Dropdowns can be used in forms on full pages, in modals, or on side panels. The dropdown component is used to filter or sort contents on a page. It is a stylized version of the select component, and can be styled as needed.

#### When not to use

> - It is best practice not to use a dropdown if there are two or less options to choose from. In this case, use a [radio button](https://www.carbondesignsystem.com/components/radio-button/code) group instead.

Reviewer: ```suggestion
- It is best practice not to use a dropdown if there are two or fewer options to choose from. In this case, use a [radio button](https://www.carbondesignsystem.com/components/radio-button/code) group instead.
```

- Do not nest dropdowns or use them to display overly complex information. Keep option selections as straight forward as possible.

> - If your user is selecting from a list of options and submitting data inside a form, use the [select component](https://www.carbondesignsystem.com/components/select/usage) instead.

Reviewer 2: This may be a misunderstanding, but this seems to conflict with a section later on under Related:

> Use a dropdown component in forms, to select multiple options at a time and to filter or sort content on a page.

Author: I know that select dropdowns are typically used when submitting data in a form, but you could also use the stylized dropdown if you wanted to right? Is there any benefit from using select over dropdown in a form? I think I just need to make a clearer argument on the uses.

- Type multiple characters in rapid succession: focus stays on the field while an option in the menu is highlighted that best matches the string of characters typed.

#### Multiple selection:

- The multiselect field receives focus if none of the options are selected in the menu.
- Once an option in the dropdown menu receives focus, users can press `Enter` to select or deselect the option.
- Users can open and close a dropdown menu by pressing `Space`.

#### Screen readers

- Labels must be announced to the screen reader on focus. Ensure the help text that appears under an input is read when an assistive technology user stops at an input using ARIA.
- VoiceOver: Users can trigger a state change by pressing `Control-Option-Space` or `Space` while the dropdown field has screen reader focus.
- JAWS: Users can trigger a state change by pressing `Space` while the dropdown field has screen reader focus.
- NVDA: Users can trigger a state change by pressing `Space` while the dropdown field has screen reader focus.

## Dropdown

Use when you can select only one option from a list at a time.

> - By default, the dropdown displays prompt text in the field when closed.

Reviewer 3: ```suggestion
- By default, the dropdown displays placeholder text in the field when closed.
```

- Clicking on a closed field opens a menu of options.

<Row>
<Column colLg={8}>

![Dropdown closed and open state.](images/dropdown-usage-9a.png)

</Column>
</Row>

> - Selecting an option from the menu closes it and the selected option text replaces the prompt text in the field and also remains as an option in place if the menu is open.

Reviewer 3: ```suggestion
- Selecting an option from the menu closes it and the selected option text replaces the placeholder text in the field and also remains as an option in place if the menu is open.
```

<Row>
<Column colLg={8}>

![Dropdown selected state.](images/dropdown-usage-9b.png)

</Column>
</Row>

## Multiselect

Use when you can select multiple options from a list or to filter information.

#### Making a selection

> - By default, the dropdown displays prompt text in the field when closed.

Reviewer 3: ```suggestion
- By default, the dropdown displays placeholder text in the field when closed.
```

- Clicking on a closed field opens a menu of options. Each option contains a checkbox input to the left of the option text.

> - The menu stays open while options are being selected. The menu closes by clicking on the field or outside of the dropdown.

Reviewer 3: ```suggestion
- The menu stays open while options are being selected. The menu closes by clicking the field or outside of the dropdown.
```

<Row>
<Column colLg={8}>

![Multiselect dropdown closed and open states.](images/dropdown-usage-10a.png)

</Column>
</Row>

> - Once options have been selected from the menu, a tag appears to the left of the text in the field containing the total number of selected options. The prompt text can change to text that better reflects what is selected.

Reviewer 3: ```suggestion
- Once options have been selected from the menu, a tag appears to the left of the text in the field containing the total number of selected options. The placeholder text can change to text that better reflects what is selected.
```

- Selected options shift to the top of the menu in alphanumeric order.
- Unlike dropdown and combobox, the menu does not close once the user makes selections. Because multiple selections are possible, the user needs to click outside of the dropdown or on the parent element to close the menu.

<Row>
<Column colLg={8}>

![Multiselect dropdown selected state.](images/dropdown-usage-10b.png)

</Column>
</Row>

#### Clear all

To clear all selected options from a list, hover over the filterable tag and click the “x” (or `close`) icon next to the value. To help with clarity, a browser tooltip appears when the user hovers over the “x” icon to indicate the click action results. If you want to unselect individual options, you can do so by unselecting the checkbox of each option.

<Row>
<Column colLg={8}>

![Multiselect dropdown clear tag interaction.](images/dropdown-usage-10c.png)

</Column>
</Row>

#### Filtering

Use filtering to narrow down a long list of options to find the option you want to select.

> - By default, the filterable multiselect dropdown displays prompt text in the field when closed.

Reviewer 3: ```suggestion
- By default, the filterable multiselect dropdown displays placeholder text in the field when closed.
```

- When hovering over the field, a text cursor appears.

<Row>
<Column colLg={8}>

![Filterable dropdown closed and open state.](images/dropdown-usage-10d.png)

</Column>
</Row>

- The menu opens by clicking anywhere in the field and you can start typing to filter the list of options.
- After typing text in the field, the close (x) icon appears to the right of the text in the field. This clears any user-inputted text. The options that start to match your entry remain in the list while other existing options are temporarily removed.

<Row>
<Column colLg={8}>

![Filterable dropdown typing and filtering open menu.](images/dropdown-usage-10e.png)

</Column>
</Row>

> - Once options have been selected from the menu, a tag appears to the left of the text in the field containing the total number of selected options. The prompt text can change to text that better reflects what is selected.

Reviewer 3: ```suggestion
- Once options have been selected from the menu, a tag appears to the left of the text in the field containing the total number of selected options. The placeholder text can change to text that better reflects what is selected.
```

- Selected options shift to the top of the menu in alphanumeric order.
- Like the default multiselect dropdown, the menu does not close once the user makes selections.

<Row>
<Column colLg={8}>

![Filterable dropdown selected state.](images/dropdown-usage-10f.png)

</Column>
</Row>

## Combobox

Allows the the user to make a selection from a predefined list of options and is typically used when there are a large amount of options to choose from.

> - By default, the combobox displays prompt text in the field when closed.

Reviewer 3: ```suggestion
- By default, the combobox displays placeholder text in the field when closed.
```

- When hovering over the field, a text cursor appears.

<Row>
<Column colLg={8}>

![Combobox closed enabled and hover states.](images/dropdown-usage-11a.png)

</Column>
</Row>

- The menu opens by clicking anywhere in the field and you can start typing to sort through the list of options.
- After typing text in the field, the close (x) icon appears to the right of the text in the field. The close icon clears any user-inputted text. The option that matches your entry receives a hover and comes into view.

<Row>
<Column colLg={8}>

![Combobox typing and option hover jumping to matched entry in the menu.](images/dropdown-usage-11b.png)

</Column>
</Row>

> - Selecting an option closes the menu and the selected option replaces the prompt text.

Reviewer 3: ```suggestion
- Selecting an option closes the menu and the selected option replaces the placeholder text.
```

 `}</ComponentVariant>
</ComponentDemo>

## Formatting

### Anatomy

Dropdowns are composed of four distinct sections—assistive text like labels or help text, a field, a menu, and options contained within the menu. Labels and help text are optional but can guide the user to make an informed decision when making a selection.

<Row>
<Column colLg={8}>

![Dropdown anatomy](images/dropdown-usage-1.png)

</Column>
</Row>

1. **Labels and help text (optional):** Assistive text to help the user make a selection.
2. **Field:** Persists when the dropdown is open or closed.
3. **Option:** A list of multiple options to select from within a menu.
4. **Menu:** Displayed as an open state and contains a list of options to select from.

### Sizing

#### Height

> There are three dropdown height sizes: Large, medium (default) and small (In code these sizes are currently called Extra large, default, and small.). Use a consistent dropdown height when it is being used alongside other form components on the same page. Supporting three different dropdown sizes gives you more flexibility when structuring layouts. When in doubt, use the default medium size height.

Reviewer 2: I bet we could have the note around the mismatch in code over on the code page if that'd help to clear this section up. Also can make an issue if that'd be helpful to rectify this.

Author: yes pleaseeee 🙏 . That would be helpful.

<Row>
<Column colLg={8}>

![Dropdown sizes](images/dropdown-usage-2.png)

</Column>
</Row>

> | Dropdown size    | Height (rem/px) | Usecase                                                                                                                                                                            |

Reviewer 2: `px` and `rem` seem to be swapped 👀 

```suggestion
| Dropdown size    | Height (px/rem) | Usecase                                                                                                                                                                            |
```

> | Dropdown size    | Height (rem/px) | Usecase                                                                                                                                                                            |

Reviewer 2: ```suggestion
| Dropdown size    | Height (rem/px) | Use case                                                                                                                                                                            |
```

Also would it be `Use case`?

<Row>
<Column colLg={8}>

![Dropdown scrolling menu option cutoff.](images/dropdown-usage-7.png)

</Column>
</Row>

### Interactions

#### Mouse

- Users trigger a `dropdown` menu to open or close by clicking the chevron icon or clicking anywhere within the field. _Note: You can only open or close a combobox by clicking the chevron icon._

<Row>
<Column colLg={8}>

![Dropdown closed and open state click targets.](images/dropdown-usage-8a.png)

</Column>
</Row>

- To select an option the user can click anywhere inside an option container.

<Row>
<Column colLg={8}>

![Dropdown menu option click target.](images/dropdown-usage-8b.png)

</Column>
</Row>

- To clear all selected options from a list in a `multiselect dropdown`, click the “x” icon next to the value inside the tag.

<Row>
<Column colLg={8}>

![Multiselect dropdown tag clear click target.](images/dropdown-usage-8c.png)

</Column>
</Row>

- To clear a selected value in a `combobox` or a filterable dropdown, click the “x” icon to the right of the field input text.

<Row>
<Column colLg={8}>

![Combobox and filterable dropdown clear click target.](images/dropdown-usage-8d.png)

</Column>
</Row>

#### Keyboard

#### Dropdown:

> - The dropdown field receives focus if none of the options are selected in the menu.

Reviewer 2: ```suggestion
- The dropdown field is the element that receives focus. All keyboard interactions happen from this element.
```

> - Users can move focus to the next option by pressing the `Down Arrow`.

Reviewer 2: ```suggestion
- Users can move the highlighted option to the next option by pressing the `Down Arrow`.
```

Focus should always stay on the field and shouldn't go between options, we end up showing a visual "highlighted" indicator as users interact with the arrow keys.

- Users can move focus to the previous option by pressing the `Up Arrow`.

> - Users can open the dropdown menu by pressing the `Spacebar`, `Return`, `Down arrow` or `Up Arrow`.

Reviewer 2: Should we use `Enter` instead of `Return`? Thinking about how WAI ARIA documents stuff over in: https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html

Author: Yes you are right, we should use `Enter` instead. 👍🏻 We are doing this for other components.

> - Users can close the dropdown menu by pressing the `Spacebar` or `Return`.

Reviewer 2: ```suggestion
- Users can close the dropdown menu by pressing the `Escape`, `Spacebar`, or `Return`
```

#### Combobox and filtering:

> - Type a character: focus moves to the next item with a name that starts with the typed character.

Reviewer 2: Similar to above, focus should stay on the field.

</Column>
</Row>

## Combobox

Allows the the user to make a selection from a predefined list of options and is typically used when there are a large amount of options to choose from.

- By default, the combobox displays prompt text in the field when closed.
- When hovering over the field, a text cursor appears.

<Row>
<Column colLg={8}>

![Combobox closed enabled and hover states.](images/dropdown-usage-11a.png)

</Column>
</Row>

- The menu opens by clicking anywhere in the field and you can start typing to sort through the list of options.
- After typing text in the field, the close (x) icon appears to the right of the text in the field. The close icon clears any user-inputted text. The option that matches your entry receives a hover and comes into view.

<Row>
<Column colLg={8}>

![Combobox typing and option hover jumping to matched entry in the menu.](images/dropdown-usage-11b.png)

</Column>
</Row>

- Selecting an option closes the menu and the selected option replaces the prompt text.

<Row>
<Column colLg={8}>

![Combobox selected state.](images/dropdown-usage-11c.png)

</Column>
</Row>

## Modifiers

#### Inline

> When placing a dropdown inline with other content use the inline modifier. If adding a visual label to an inline dropdown it should appear inline to the left of the dropdown. If there is no visual label present we recommend applying an ARIA label to the inline dropdown. _Note: Inline is only a modifier for dropdown and does not have filtering functionality._

Reviewer 2: Might be helpful for us to be more forceful with the language, from:

> If there is no visual label present we recommend applying an ARIA label to the inline dropdown.

To:

> If there is no visual label present you must supply an appropriate accessibility label to the inline dropdown

![Dropdown disabled state.](images/dropdown-usage-15.png)

</Column>
</Row>

## Related

#### Dropdown versus Select

Dropdown and select components have functionality and style differences.

- Use a dropdown component in forms, to select multiple options at a time and to filter or sort content on a page.
- Use the select component in forms where users are selecting from a list of options and submitting data. The select component does not have filtering or multiselect functionality.
- The underlying code of a dropdown component is styled to match the design system, while the select component’s appearance is determined by the browser being used. \*\*

#### Dropdown versus Combobox

While the dropdown and combobox look similar they have different functions.

- With a dropdown list, the selected option is always visible, and the other options are visible by clicking and opening the list.
- A combobox is a combination of a standard list box or a dropdown list and an editable text box and users are able to enter a value that isn't in the list.

## References

Angie Li, [Dropdown: Design Guidelines](https://www.nngroup.com/articles/drop-down-menus/) (Nielsen Norman Group, 2017)

> -[Checkbox](https://www.carbondesignsystem.com/components/checkbox/code)

Reviewer 2: ```suggestion
- [Checkbox](https://www.carbondesignsystem.com/components/checkbox/code)
```
